corehq/apps/locations/views.py

```diff
@@ -342,14 +344,33 @@ def mk_loctype(name, parent_type, administrative, has_user,
             pk = loc_type['pk']
             if not _is_fake_pk(pk):
                 pks.append(loc_type['pk'])
-
-        names = [lt['name'] for lt in loc_types]
+            payload_loc_type_name_by_pk[loc_type['pk']] = loc_type['name']
+            if loc_type['code']:
+                payload_loc_type_code_by_pk[loc_type['pk']] = loc_type['code']
+        names = payload_loc_type_name_by_pk.values()
         names_are_unique = len(names) == len(set(names))
-        codes = [lt['code'] for lt in loc_types if lt['code']]
+        codes = payload_loc_type_code_by_pk.values()
         codes_are_unique = len(codes) == len(set(codes))
         if not names_are_unique or not codes_are_unique:
             raise LocationConsistencyError("'name' and 'code' are supposed to be unique")
 
+        current_location_types = LocationType.objects.by_domain(request.domain)
+        for location_type in current_location_types:
+            if location_type.pk in payload_loc_type_name_by_pk:
+                # to check if the name/code was swapped with another location by confirming if
+                # either name/code has changed but the current name is still present in the names/codes passed
+                if (
+                        (location_type.name != payload_loc_type_name_by_pk.get(location_type.pk) and
+                         location_type.name in names) or
+                        (location_type.code != payload_loc_type_code_by_pk[location_type.pk] and
+                         location_type.code in codes)
+                ):
+                    messages.error(request, LocationConsistencyError(_(
+                        "Looks like you are assigning a location name/code to a different location "
+                        "in the same request. Please do this in two separate updates by using a "
+                        "temporary name to free up the name/code to be re-assigned."))
+                    )
```

Comment:
I think we should be able to work around this in some way - would wrapping all the saves in a single transaction allow this to go through as-is?

Reply:
@esoergel i don't think that would help since the save would still happen one by one and the unique clause would fail?

Reply:
Dang, I was hoping postgres would defer unique checking until the transaction is finished, but that doesn't appear to be the case (just tried it out in a shell session).

Another option is to assign one of the location types a temporary name (say, a UUID), then make a second pass and save it again with the real name.  It's an extra round-trip to the db, but we expect there to be a very small number of location types on any given domain, and it beats making the user do that manually.

Reply:
Huh, okay.  I do think that would be better behavior, but I haven't tried it to see how complex that is to implement.  I'll leave it up to your judgement on whether it's worth it.

Reply:
I had similar thoughts as @esoergel but I guess if it does introduce a lot of complexity then fine to leave as is for now. If this comes up again we can do it then.

Reply:
Thank you.